### PR TITLE
Simplify Beater interface

### PR DESCRIPTION
**Pull request by urso**
pass Beater constructor to beat.Run + simplify Beater interface to only require
Run and Stop methods.

### CODE REVIEW

**Comment by urso** [comment]
@ruflin any tip how to fix the beat generator test? It seems the generator uses master branch, but not feature branch. Maybe fix generator after merging the PR?

File: filebeat/config/config.go [hunk 1]
```diff
@@ -98,13 +99,13 @@ func mergeConfigFiles(configFiles []string, config *Config) error {
 }
 
 // Fetches and merges all config files given by configDir. All are put into one config object
-func (config *Config) FetchConfigs() {
+func (config *Config) FetchConfigs() error {
```
Comment (ruflin) [inline filebeat/config/config.go]: Thanks for the cleanup here. This part could probably go into a separate PR to keep the changes in this one to a minimum.

File: filebeat/publish/publish.go [hunk 2]
```diff
@@ -18,6 +18,7 @@ type LogPublisher interface {
 }
 
 type syncLogPublisher struct {
+	pub     publisher.Publisher
```
Comment (ruflin) [inline filebeat/publish/publish.go]: Changes to the publisher could also go into a separate filebeat cleanup PR?

File: generate/beat/{{cookiecutter.beat}}/beater/{{cookiecutter.beat}}.go [hunk 3]
```diff
@@ -13,54 +13,32 @@ import (
...
-
-	return nil
-}
+/// *** Beater interface methods ***///
 
 func (bt *{{cookiecutter.beat|capitalize}}) Run(b *beat.Beat) error {
 	logp.Info("{{cookiecutter.beat}} is running! Hit CTRL-C to stop it.")
 
-	ticker := time.NewTicker(bt.period)
+	bt.client = b.Publisher.Connect()
```
Comment (ruflin) [inline generate/beat/{{cookiecutter.beat}}/beater/{{cookiecutter.beat}}.go]: An idea when I saw this: Should we require each beat to take care of the connec on Run or should we already to it before in libbeat?

File: generate/beat/{{cookiecutter.beat}}/config/config.go [hunk 4]
```diff
@@ -3,10 +3,18 @@
...
 
+import "time"
+
 type Config struct {
 	{{cookiecutter.beat|capitalize}} {{cookiecutter.beat|capitalize}}Config
 }
 
 type {{cookiecutter.beat|capitalize}}Config struct {
-	Period string `config:"period"`
+	Period time.Duration `config:"period"`
```
Comment (ruflin) [inline generate/beat/{{cookiecutter.beat}}/config/config.go]: +1 on these changes. Could go into a cookiecutter cleanup PR?

File: libbeat/beat/beat.go [hunk 5]
```diff
@@ -150,70 +139,113 @@ func init() {
...
-
-	return err
+	return handleFlags(b)
 }
 
 // config reads the configuration file from disk, parses the common options
 // defined in BeatConfig, initializes logging, and set GOMAXPROCS if defined
 // in the config. Lastly it invokes the Config method implemented by the beat.
-func (bc *instance) config() error {
+func (b *Beat) configure() error {
```
Comment (ruflin) [inline libbeat/beat/beat.go]: I would prefer to keep the name config or if we change it, also comment must be changed.

File: libbeat/publisher/client.go [hunk 6]
```diff
@@ -62,12 +62,12 @@ type Client interface {
...
 	canceler *op.Canceler
 
-	publisher           *Publisher
+	publisher           *BeatPublisher
 	beatMeta            common.MapStr        // Beat metadata that is added to all events.
 	globalEventMetadata common.EventMetadata // Fields and tags that are added to all events.
 }
 
-func newClient(pub *Publisher) *client {
+func newClient(pub *BeatPublisher) *client {
```
Comment (ruflin) [inline libbeat/publisher/client.go]: Renaming *BeatPublisher could go into a libbeat cleanup PR.

File: packetbeat/beater/packetbeat.go [hunk 7]
```diff
@@ -27,7 +27,7 @@ import (
 
 // Beater object. Contains all objects needed to run the beat
 type Packetbeat struct {
-	PbConfig    config.Config
+	Config      config.Config
```
Comment (ruflin) [inline packetbeat/beater/packetbeat.go]: Packetbeat cleanup PR

File: packetbeat/publish/publish.go [hunk 8]
```diff
@@ -19,9 +20,13 @@ type Flows interface {
 }
 
 type PacketbeatPublisher struct {
-	pub    *publisher.Publisher
+	pub    publisher.Publisher
```
Comment (ruflin) [inline packetbeat/publish/publish.go]: Packetbeat cleanup

File: winlogbeat/config/config.go [hunk 9]
```diff
@@ -31,6 +31,14 @@ type Settings struct {
 	Raw        map[string]interface{} `config:",inline"`
 }
 
+var (
+	DefaultSettings = Settings{
```
Comment (ruflin) [inline winlogbeat/config/config.go]: Windlogbeat cleanup PR

File: libbeat/beat/beat.go [hunk 10]
```diff
@@ -52,15 +52,7 @@ import (
...
-	printVersion = flag.Bool("version", false, "Print the version and exit")
-)
-
-var debugf = logp.MakeDebug("beat")
-
-// GracefulExit is an error that signals to exit with a code of 0.
-var GracefulExit = errors.New("graceful exit")
-
+/*
 // Beater is the interface that must be implemented every Beat. The full
```
Comment (ruflin) [inline libbeat/beat/beat.go]: The docs below must be udpated.

**Comment by ruflin** [comment]
+1 on these changes. In general LGTM. 

For my understanding:
* Config(), Setup(), Cleanup() interface methods were removed
* Config is now part of New
* Setup is part or Run
* Cleanup is part of Run as defer or can be part of Stop

So migrating from the old to the new interfaces in a simplified way is to keep the methods like Setup but call them manually in the "right place"?

This changes make the beat interface very similar to the MetricSet interface which I quite like.

The docs must be updated with the new interfaces. As this change breaks all existing beats, we must definitively document it briefly in the changelog and potentially link to a "migration guide" in the dev guide.

I left a few comments similar to "cleanup PR x". The reasons for this is that most of the changes in this PR are actually not related to the interface changes but some cleanup which is nice. But as most of the beats developers have to migrate their beat I would prefer to have a PR with only the changes from the migration inside. This makes it much easier for people to migrate I hope. I'm aware of the separate commits but I prefer to point people to a PR with the changes. Could you pick some of the commits above and open a "Cleanup PR"?

About beat-generator: Unfortunately yes. I implemented a workaround for the metricset-generator which I should also apply to beat generator. Let me see if I can do that today.

**Comment by ruflin** [comment]
@urso generator problem should be fixed by https://github.com/elastic/beats/pull/2006

File: generate/beat/{{cookiecutter.beat}}/beater/{{cookiecutter.beat}}.go (hunk 3 above)
Comment (urso) [inline generate/beat/{{cookiecutter.beat}}/beater/{{cookiecutter.beat}}.go]: touch one. But it's perfectly valid for a beat connect multiple times. So i'd keep it as is for now. We should only make it safe to connect to publisher pipeline in constructor, that is, libbeat launch method will kill/break connection on shutdown automatically (to be done in future PR).

File: libbeat/beat/beat.go (hunk 5 above)
Comment (urso) [inline libbeat/beat/beat.go]: I named it configure, not to be confused with 'Config' field in Beat.

**Comment by urso** [comment]
> For my understanding:
- Config(), Setup(), Cleanup() interface methods were removed
- Config is now part of New
- Setup is part or Run
- Cleanup is part of Run as defer or can be part of Stop

Yes and no:
- Config(), Setup(), Cleanup() have been removed
- Config must be done in New
- Setup depends on beat. Implementation of Setup has been a little inconsistent among beat. Some beats doing too much work in Setup require to call setup in Run, others require setup to be run in New.
- Cleanup should be part of Run as defer. Like done for filebeat more recently. Stop is supposed to break the run loop.

File: libbeat/beat/beat.go (hunk 10 above)
Comment (urso) [inline libbeat/beat/beat.go]: yes, just commented out original interface. still needs to be cleaned up here. `in progress` flag ;)

File: filebeat/config/config.go [hunk 11]
```diff
@@ -117,14 +118,20 @@ func (config *Config) FetchConfigs() {
 
 	if err != nil {
 		log.Fatal("Could not use config_dir of: ", configDir, err)
```
Comment (ruflin) [inline filebeat/config/config.go]: Instead of logging the message here, we could just return the error. Can be done in a later PR.

File: filebeat/publish/publish.go [hunk 12]
```diff
@@ -127,8 +127,8 @@ func (p *syncLogPublisher) Start() {
 }
 
 func (p *syncLogPublisher) Stop() {
-	close(p.done)
 	p.client.Close()
+	close(p.done)
```
Comment (ruflin) [inline filebeat/publish/publish.go]: I'm curious if that could have led to some issues previously?

File: generate/beat/{{cookiecutter.beat}}/beater/{{cookiecutter.beat}}.go [hunk 13]
```diff
@@ -13,42 +13,29 @@ import (
 )
 
 type {{cookiecutter.beat|capitalize}} struct {
-	config config.Config
-	done   chan struct{}
-	client publisher.Client
+	done       chan struct{}
```
Comment (ruflin) [inline generate/beat/{{cookiecutter.beat}}/beater/{{cookiecutter.beat}}.go]: It could be nice to have the done channel also in under Beat.Done so simple beats could reuse it without having to create or close one themself.

File: generate/beat/{{cookiecutter.beat}}/beater/{{cookiecutter.beat}}.go [hunk 14]
```diff
@@ -13,42 +13,29 @@ import (
...
+	config     config.Config
```
Comment (ruflin) [inline generate/beat/{{cookiecutter.beat}}/beater/{{cookiecutter.beat}}.go]: BTW: I assume the build is going to complain here about `go fmt` ...

File: generate/beat/{{cookiecutter.beat}}/beater/{{cookiecutter.beat}}.go [hunk 15]
```diff
@@ -13,42 +13,29 @@ import (
...
-
-func (bt *{{cookiecutter.beat|capitalize}}) Config(b *beat.Beat) error {
 
-	bt.config = config.DefaultConfig
-
-	// Load beater config
-	err := b.RawConfig.Unpack(&bt.config)
-	if err != nil {
-		return fmt.Errorf("Error reading config file: %v", err)
+	bt := &{{cookiecutter.beat|capitalize}}{
```
Comment (ruflin) [inline generate/beat/{{cookiecutter.beat}}/beater/{{cookiecutter.beat}}.go]: Long term we could also go here with a generic approach as we did for MetricSet, calling it Beater here. But it is currently nice to have the name.

File: libbeat/beat/beat.go [hunk 16]
```diff
@@ -52,53 +52,40 @@ import (
...
+	Run(b *Beat) error
+
+	// Stop is invoked to signal that the Run method should finish its execution.
+	// It will be invoked at most once.
+	Stop()
 }
 
+// Creator initializes and configures a new Beater instance used to execute
+// the beat its run-loop.
+type Creator func(*Beat, *common.Config) (Beater, error)
```
Comment (ruflin) [inline libbeat/beat/beat.go]: Not 100% sure about the name.

File: libbeat/beat/beat.go [hunk 17]
```diff
@@ -141,65 +124,114 @@ func init() {
...
+
+func (b *Beat) launch(bt Creator) error {
+	if err := b.handleFlags(); err != nil {
+		return err
+	}
+
+	svc.BeforeRun()
+	defer svc.Cleanup()
+
+	if err := b.configure(); err != nil {
```
Comment (ruflin) [inline libbeat/beat/beat.go]: We loose some flexibilty here in that no special tasks can be execute by a Beater between configuring the beat and running the beat. This was previously possible. But as the refactoring just worked, I assume this possibility was not needed.

File: libbeat/beat/beat.go [hunk 18]
```diff
@@ -141,65 +124,114 @@ func init() {
...
+		return fmt.Errorf("error initializing processors: %v", err)
+	}
+
+	debugf("Initializing output plugins")
+	publisher, err := publisher.New(b.Name, b.Config.Output, b.Config.Shipper, processors)
+	if err != nil {
+		return fmt.Errorf("error initializing publisher: %v", err)
+	}
+
+	// TODO: some beats race on shutdown with publisher.Stop -> do not call Stop yet,
```
Comment (ruflin) [inline libbeat/beat/beat.go]: What happens to this comment?

File: filebeat/publish/publish.go (hunk 12 above)
Comment (urso) [inline filebeat/publish/publish.go]: not really. libbeat is not checking connections being closed yet (disabled due to races). There is still a chance events are published besides being shutdown.

File: generate/beat/{{cookiecutter.beat}}/beater/{{cookiecutter.beat}}.go (hunk 13 above)
Comment (urso) [inline generate/beat/{{cookiecutter.beat}}/beater/{{cookiecutter.beat}}.go]: hm, with libbeat providing Done and maybe even closing client on shutdown, we still need a stop method?

This might even be a bigger change in API.

for example, if I need to unblock anything once done closes I can write:

```
go func(){
    <-b.Done
    bt.UnblockWhatever()
}()
```

File: packetbeat/beater/packetbeat.go [hunk 19]
```diff
@@ -135,16 +126,76 @@ func (pb *Packetbeat) Setup(b *beat.Beat) error {
...
+
+func (pb *Packetbeat) Run(b *beat.Beat) error {
+	defer func() {
+		if service.WithMemProfile() {
+			logp.Debug("main", "Waiting for streams and transactions to expire...")
+			time.Sleep(time.Duration(float64(protos.DefaultTransactionExpiration) * 1.2))
+			logp.Debug("main", "Streams and transactions should all be expired now.")
+		}
+
+		// TODO:
```
Comment (ruflin) [inline packetbeat/beater/packetbeat.go]: What will happen to this one?

File: generate/beat/{{cookiecutter.beat}}/beater/{{cookiecutter.beat}}.go (hunk 14 above)
Comment (urso) [inline generate/beat/{{cookiecutter.beat}}/beater/{{cookiecutter.beat}}.go]: build is all greenish. consider adding a 'go fmt' check. Maybe it's some good practice to call `go fmt` as post-processing step to ensure all files generated are in good shape.

File: libbeat/beat/beat.go (hunk 16 above)
Comment (urso) [inline libbeat/beat/beat.go]: The full name is `beat.Creator` ;)

File: packetbeat/beater/packetbeat.go [hunk 20]
```diff
@@ -202,64 +253,3 @@ func (pb *Packetbeat) makeWorkerFactory(filter string) sniffer.WorkerFactory {
...
-
-func (pb *Packetbeat) Cleanup(b *beat.Beat) error {
-
-	if service.WithMemProfile() {
-		logp.Debug("main", "Waiting for streams and transactions to expire...")
-		time.Sleep(time.Duration(float64(protos.DefaultTransactionExpiration) * 1.2))
-		logp.Debug("main", "Streams and transactions should all be expired now.")
-	}
-
-	// TODO:
```
Comment (ruflin) [inline packetbeat/beater/packetbeat.go]: I see, this is only copy/paste :-)

File: libbeat/beat/beat.go (hunk 17 above)
Comment (urso) [inline libbeat/beat/beat.go]: I think it's good we're loosing this `flexibility`. Setup used to be somewhere in the middle between configure and run, inconsistently used by multiple beats as such.

File: packetbeat/publish/publish.go [hunk 21]
```diff
@@ -99,8 +99,8 @@ func (t *PacketbeatPublisher) Start() {
 }
 
 func (t *PacketbeatPublisher) Stop() {
```
Comment (ruflin) [inline packetbeat/publish/publish.go]: ok, there is a separate packetbeat publisher

**Comment by ruflin** [comment]
LGTM

File: libbeat/beat/beat.go (hunk 18 above)
Comment (urso) [inline libbeat/beat/beat.go]: looks ok to me. comment explains why `defer publisher.Stop()` is still commented out.

File: generate/beat/{{cookiecutter.beat}}/beater/{{cookiecutter.beat}}.go (hunk 13 above)
Comment (ruflin) [inline generate/beat/{{cookiecutter.beat}}/beater/{{cookiecutter.beat}}.go]: Good question. Not sure TBH. We should not only look at our beats but also at the community beats on how they use the interfaces.

File: generate/beat/{{cookiecutter.beat}}/beater/{{cookiecutter.beat}}.go (hunk 14 above)
Comment (ruflin) [inline generate/beat/{{cookiecutter.beat}}/beater/{{cookiecutter.beat}}.go]: not so easy here because of the template variables. I just remember fixing that manually, but if tests are green, all good. because it checks it as part of the build.

File: libbeat/beat/beat.go (hunk 16 above)
Comment (ruflin) [inline libbeat/beat/beat.go]: I currently don't have a better suggestion :-)

File: packetbeat/beater/packetbeat.go (hunk 19 above)
Comment (urso) [inline packetbeat/beater/packetbeat.go]: no idea yet. Just copy'n pasted from original location. Gotta check if it can be removed (another PR).

File: packetbeat/beater/packetbeat.go [hunk 22]
```diff
@@ -135,16 +126,76 @@ func (pb *Packetbeat) Setup(b *beat.Beat) error {
...
+	}
+
 	return nil
 }
 
+// Called by the Beat stop function
+func (pb *Packetbeat) Stop() {
+	logp.Info("Packetbeat send stop signal")
+	pb.Sniff.Stop()
+	pb.Pub.Stop()
```
Comment (urso) [inline packetbeat/beater/packetbeat.go]: this is a PacketbeatPublisher worker having 'Stop' method.

**Comment by ruflin** [comment]
@urso I seems like the appveyor tests indicated that something changed in the exit behaviour of winlogbeat.

File: libbeat/beat/beat.go [hunk 23]
```diff
@@ -52,53 +52,40 @@ import (
 	"github.com/satori/go.uuid"
 )
```
Comment (andrewkroh) [inline libbeat/beat/beat.go]: ^ The package level comment is out-of-date.

File: libbeat/beat/beat.go [hunk 24]
```diff
@@ -141,65 +124,114 @@ func init() {
...
+// instance.
+func Run(name, version string, bt Creator) error {
+	return handleError(newBeat(name, version).launch(bt))
+}
+
+// newBeat creates a new beat instance
+func newBeat(name, version string) *Beat {
+	return &Beat{
+		Name:    name,
+		Version: version,
```
Comment (andrewkroh) [inline libbeat/beat/beat.go]: For our beats, `version` is empty, so I think this should check for empty and then set version to `defaultBeatVersion`. At startup our beats seem to be printing `Version:` but no number.

File: winlogbeat/beater/winlogbeat.go [hunk 25]
```diff
@@ -114,8 +106,8 @@ func (eb *Winlogbeat) Setup(b *beat.Beat) error {
 
 	// Create the event logs. This will validate the event log specific
 	// configuration.
-	eb.eventLogs = make([]eventlog.EventLog, 0, len(eb.config.Winlogbeat.EventLogs))
-	for _, config := range eb.config.Winlogbeat.EventLogs {
+	eb.eventLogs = make([]eventlog.EventLog, 0, len(config.EventLogs))
+	for _, config := range config.EventLogs {
 		eventLog, err := eventlog.New(config)
```
Comment (andrewkroh) [inline winlogbeat/beater/winlogbeat.go]: There is a lot of useful configuration validation performed as part of this `eventlog.New` method. It would be best if most of what is in this `Setup` method was invoked as part of `-configtest` (with the exception of the part where it launches an http server).

File: generate/beat/{{cookiecutter.beat}}/beater/{{cookiecutter.beat}}.go (hunk 14 above)
Comment (urso) [inline generate/beat/{{cookiecutter.beat}}/beater/{{cookiecutter.beat}}.go]: What I have in mind is running `go fmt` via cookiecutter post-generate hook: http://cookiecutter.readthedocs.io/en/latest/advanced/hooks.html

File: libbeat/beat/beat.go (hunk 16 above)
Comment (urso) [inline libbeat/beat/beat.go]: Creator, Constructor, Builder, Maker, Factory, Initializor.

File: libbeat/beat/beat.go (hunk 23 above)
Comment (urso) [inline libbeat/beat/beat.go]: Thanks. Updated.

File: libbeat/beat/beat.go (hunk 24 above)
Comment (urso) [inline libbeat/beat/beat.go]: Thanks. fixed.

File: winlogbeat/beater/winlogbeat.go (hunk 25 above)
Comment (urso) [inline winlogbeat/beater/winlogbeat.go]: moved eventlog.New into `init` method. Not sure about adding checkpoint to `init`. It checks the checkpoint file is available, can be parsed, can be written to + starts some worker. All of this required by `-configtest`?

Comment (andrewkroh) [inline winlogbeat/beater/winlogbeat.go]: Good point about checkpoint. Validating the checkpoint file isn't strictly required to be done in `-configtest` so this doesn't need to be done in `New` since it starts the worker.

File: libbeat/beat/beat.go (hunk 16 above)
Comment (ruflin) [inline libbeat/beat/beat.go]: Lets keep it as it is :-)

**Comment by ruflin** [comment]
LGTM
